File: packages/backend/survey/src/survey-repository.ts
/**
 * Survey Repository Interfaces
 *
 * Defines the data access contracts for survey operations.
 * Implementations can use Prisma, in-memory stores, or other backends.
 *
 * Requirements: 23.1, 23.2, 23.3, 23.4, 23.5
 */
import type { PaginationOptions, PaginatedResult } from '@proctira/common';
import type {
  QuestionType,
  SurveyStatus,
  CompletionStatus,
  DropdownOption,
  TableColumn,
} from './schemas.js';

// ─── Survey Entity ───────────────────────────────────────────────────────────

/**
 * Survey entity as stored in the database.
 */
export interface SurveyEntity {
  id: string;
  tenantId: string;
  name: string;
  description: string | null;
  status: SurveyStatus;
  academicPeriodId: string | null;
  startDate: string | null;
  endDate: string | null;
  questions: QuestionEntity[];
  createdAt: Date;
  updatedAt: Date;
}

/**
 * Question entity embedded within a survey.
 */
export interface QuestionEntity {
  id: string;
  label: string;
  type: QuestionType;
  required: boolean;
  order: number;
  options?: DropdownOption[];
  columns?: TableColumn[];
  repeaterFields?: RepeaterFieldEntity[];
  validation?: QuestionValidation;
}

export interface RepeaterFieldEntity {
  label: string;
  type: 'text' | 'number' | 'date' | 'dropdown';
  required: boolean;
  options?: DropdownOption[];
}

export interface QuestionValidation {
  min?: number;
  max?: number;
  minLength?: number;
  maxLength?: number;
  pattern?: string;
}

/**
 * Filter options for listing surveys.
 */
export interface SurveyFilter {
  status?: SurveyStatus;
  search?: string;
}

/**
 * Repository interface for survey data access.
 */
export interface SurveyRepository {
  /** Create a new survey */
  create(data: Omit<SurveyEntity, 'createdAt' | 'updatedAt'>): Promise<SurveyEntity>;

  /** Update an existing survey */
  update(id: string, tenantId: string, data: Partial<Omit<SurveyEntity, 'id' | 'tenantId' | 'createdAt' | 'updatedAt'>>): Promise<SurveyEntity | null>;

  /** Find a survey by ID within a tenant */
  findById(id: string, tenantId: string): Promise<SurveyEntity | null>;

  /** Find a survey by name within a tenant */
  findByName(name: string, tenantId: string): Promise<SurveyEntity | null>;

  /** Delete a survey */
  delete(id: string, tenantId: string): Promise<boolean>;

  /** List surveys with pagination and filtering */
  list(
    tenantId: string,
    filter: SurveyFilter,
    pagination: PaginationOptions,
  ): Promise<PaginatedResult<SurveyEntity>>;
}

// ─── Distribution Record Entity ──────────────────────────────────────────────

/**
 * Distribution record tracking survey assignment to an institution.
 * Requirement 23.2, 23.4
 */
export interface DistributionRecordEntity {
  id: string;
  tenantId: string;
  surveyId: string;
  institutionId: string;
  status: CompletionStatus;
  dueDate: string | null;
  submittedAt: string | null;
  remindersSent: number;
  reminderDays: number[];
  createdAt: Date;
  updatedAt: Date;
}

/**
 * Repository interface for distribution record data access.
 */
export interface DistributionRepository {
  /** Create distribution records for multiple institutions */
  createMany(records: Omit<DistributionRecordEntity, 'createdAt' | 'updatedAt'>[]): Promise<DistributionRecordEntity[]>;

  /** Find distribution record for a specific survey and institution */
  findBySurveyAndInstitution(tenantId: string, surveyId: string, institutionId: string): Promise<DistributionRecordEntity | null>;

  /** Find all distribution records for a survey */
  findBySurvey(tenantId: string, surveyId: string): Promise<DistributionRecordEntity[]>;

  /** Find incomplete distribution records for a survey */
  findIncompleteBySurvey(tenantId: string, surveyId: string): Promise<DistributionRecordEntity[]>;

  /** Update a distribution record */
  update(id: string, tenantId: string, data: Partial<Pick<DistributionRecordEntity, 'status' | 'submittedAt' | 'remindersSent'>>): Promise<DistributionRecordEntity | null>;

  /** Count distribution records by status for a survey */
  countByStatus(tenantId: string, surveyId: string): Promise<Record<CompletionStatus, number>>;
}

// ─── Submission Entity ───────────────────────────────────────────────────────

/**
 * Survey submission entity storing responses from an institution.
 * Requirement 23.3
 */
export interface SubmissionEntity {
  id: string;
  tenantId: string;
  surveyId: string;
  institutionId: string;
  answers: AnswerEntity[];
  submittedAt: Date;
  createdAt: Date;
}

export interface AnswerEntity {
  questionId: string;
  value: unknown;
}

/**
 * Repository interface for submission data access.
 */
export interface SubmissionRepository {
  /** Create a new submission */
  create(data: Omit<SubmissionEntity, 'createdAt'>): Promise<SubmissionEntity>;

  /** Find submissions for a survey */
  findBySurvey(tenantId: string, surveyId: string): Promise<SubmissionEntity[]>;

  /** Find submission for a specific survey and institution */
  findBySurveyAndInstitution(tenantId: string, surveyId: string, institutionId: string): Promise<SubmissionEntity | null>;
}

// ─── Institution Lookup Interface ────────────────────────────────────────────

/**
 * Interface for looking up institutions based on distribution filters.
 * Requirement 23.2: Distribute based on area, type, classification.
 */
export interface InstitutionLookup {
  /** Find institution IDs matching the given filters */
  findByFilters(
    tenantId: string,
    filters: {
      areaIds?: string[];
      institutionTypeIds?: string[];
      classificationIds?: string[];
    },
  ): Promise<string[]>;

  /** Get institution metadata for aggregation grouping */
  getMetadata(tenantId: string, institutionIds: string[]): Promise<InstitutionMetadata[]>;
}

export interface InstitutionMetadata {
  id: string;
  name: string;
  areaId: string;
  areaName: string;
  typeId: string;
  typeName: string;
}

File: apps/mobile/lib/features/survey/data/survey_repository.dart
import 'package:dio/dio.dart';

class SurveyQuestion {
  const SurveyQuestion({
    required this.label,
    required this.type,
    required this.order,
    this.id,
    this.required = false,
    this.options = const <({String label, String value})>[],
  });

  final String? id;
  final String label;
  final String type;
  final int order;
  final bool required;
  final List<({String label, String value})> options;

  factory SurveyQuestion.fromJson(Map<String, dynamic> json) {
    final List<dynamic>? rawOptions = json['options'] as List<dynamic>?;
    return SurveyQuestion(
      id: json['id'] as String?,
      label: json['label'] as String? ?? '',
      type: json['type'] as String? ?? 'text',
      order: json['order'] as int? ?? 0,
      required: json['required'] as bool? ?? false,
      options: rawOptions
              ?.whereType<Map>()
              .map(
                (Map e) => (
                  label: e['label'] as String? ?? '',
                  value: e['value'] as String? ?? '',
                ),
              )
              .toList(growable: false) ??
          const <({String label, String value})>[],
    );
  }
}

class Survey {
  const Survey({
    required this.id,
    required this.name,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.description,
    this.startDate,
    this.endDate,
    this.questions = const <SurveyQuestion>[],
  });

  final String id;
  final String name;
  final String status;
  final String createdAt;
  final String updatedAt;
  final String? description;
  final String? startDate;
  final String? endDate;
  final List<SurveyQuestion> questions;

  factory Survey.fromJson(Map<String, dynamic> json) {
    final List<dynamic>? rawQuestions = json['questions'] as List<dynamic>?;
    return Survey(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      status: json['status'] as String? ?? 'draft',
      createdAt: json['createdAt'] as String? ?? '',
      updatedAt: json['updatedAt'] as String? ?? '',
      description: json['description'] as String?,
      startDate: json['startDate'] as String?,
      endDate: json['endDate'] as String?,
      questions: rawQuestions
              ?.whereType<Map>()
              .map(
                (Map e) =>
                    SurveyQuestion.fromJson(Map<String, dynamic>.from(e)),
              )
              .toList(growable: false) ??
          const <SurveyQuestion>[],
    );
  }
}

/// Dio client for `/api/v1/surveys`.
class SurveyRepository {
  SurveyRepository({required Dio dio}) : _dio = dio;

  final Dio _dio;

  Future<List<Survey>> listSurveys({String? status}) async {
    try {
      final Response<dynamic> response = await _dio.get(
        '/api/v1/surveys',
        queryParameters: <String, dynamic>{
          'pageSize': 100,
          'status': ?status,
        },
      );
      return _unwrapList(response.data)
          .map(Survey.fromJson)
          .toList(growable: false);
    } on DioException {
      return const <Survey>[];
    }
  }

  Future<Survey?> getSurvey(String id) async {
    try {
      final Response<dynamic> response = await _dio.get('/api/v1/surveys/$id');
      final Object? raw = response.data;
      if (raw is! Map) return null;
      return Survey.fromJson(Map<String, dynamic>.from(raw));
    } on DioException {
      return null;
    }
  }

  Future<void> submitResponse({
    required String surveyId,
    required String institutionId,
    required List<Map<String, dynamic>> answers,
  }) async {
    await _dio.post(
      '/api/v1/surveys/submit',
      data: <String, dynamic>{
        'surveyId': surveyId,
        'institutionId': institutionId,
        'answers': answers,
      },
    );
  }

  List<Map<String, dynamic>> _unwrapList(dynamic payload) {
    if (payload == null) return const <Map<String, dynamic>>[];
    if (payload is List) {
      return payload
          .whereType<Map>()
          .map((Map e) => Map<String, dynamic>.from(e))
          .toList(growable: false);
    }
    if (payload is Map && payload['data'] is List) {
      return (payload['data'] as List)
          .whereType<Map>()
          .map((Map e) => Map<String, dynamic>.from(e))
          .toList(growable: false);
    }
    return const <Map<String, dynamic>>[];
  }
}
